### ISSUE TITLE
dom-if template evaluates data once even if the condition is false

### ISSUE THREAD

**Issue by namiwang** [Polymer/polymer#1716]
Basically this is a duplicated issue of #1663 , which is closed but I think the issue is still not fixed. Hope it doesn't bother you.

I just make a more proper demo [here](http://jsbin.com/dusaseteze/1/edit?html,console,output) to show the bug.

Re-pro step:
1. click set apple
2. click set orange

What's wrong:
- As the console output said, the `parseAppleSpecies()` have been called twice, which should be only once.

**Comment by sorvell** [comment]
@kevinpschaaf We probably want to consider turning off bindings to/from dom-if when its `if` is false.

**Comment by kevinpschaaf** [comment]
You are seeing `parseAppleSpecies` being run for two reasons:

1. By default, `dom-if` does not discard previously stamped nodes when `if` becomes false again, for better performance when it becomes `true`; in this case the nodes are only hidden, not removed & unbound, and will continue reacting to bindings.

2. You can defeat this behavior by setting the `restamp` attribute on the `dom-if`, which will cause the nodes to be removed & discarded, then re-created later.  However, you'll still see `parseAppleSpecies` run even with `restamp` due to the fact that `dom-if` is "just" a custom element in Polymer's eyes, so the binding to `if` has no special significance, and the binding to `parseAppleSpecies` just happens to run before the binding to the `if` that removes it (in general the order is not to be relied on).  We may consider doing something special for the ordering of `dom-if` since it's hard to argue that this is unexpected.

@nanowang Aside from the fact that it was unexpected, are you hitting a scenario where running the computed function is actually causing a problem?  Or was it just a curiosity or concern about the performance implications?  If you have a specific scenario where this is blocking you, that would be useful for us to consider in prioritizing if/when to fix this.

Thanks for the repro case.

**Comment by nanowang** [comment]
@kevinpschaaf 

Hi, thanks for your reply and explanation.

Here's my scenario: I'm developing an engine for adventure games using Polymer, it has got some complicated GUI boxes, like an inventory box, or a text-box displaying the conversation, which is related to this issue.

In detail, the content(which is called `node` in my engine) of one `conversation-box` could be either a line said by a character from the script, or narration, or some dialog options, etc.

So I give a `type` to each `node`. Since I can't do something like `if={{node.type == 'line'}}`, I also have to set some flags like `node.typeIsNarration`, `node.typeIsLine`, etc. for every possible type, which is truly annoying but is another issue. :(

To render the conversation-box, I've got a template like this:
```haml
%dom-module#aozora-conversation-box
  %link{rel: 'import', type: 'css', href: 'aozora-conversation-box.css'}
  %template
    %template{is: 'dom-if', if: '{{node.typeIsLine}}'}
      .node.line
        %p.role {{ roleNameOfLine(node) }} <== here's the problem
        %p {{ node.text }}

    %template{is: 'dom-if', if: '{{node.typeIsNarration}}'}
      .node.narrate
        %p {{ node.text }}

    %template{is: 'dom-if', if: '{{node.typeIsOptions}}'}
      .node.options
        %template{is: 'dom-repeat', items: '[[node.options]]', as: 'option'}
          .option
            %p{'on-tap' => 'onTapOnOption'} {{ option.text }}
```

The problem is, when the current `node` is changing from a `line` into a `narration`, the method `roleNameOfLine()` would be unexpectedly called, but the method is designed for only one type of node.

I guess this is a common issue, since a container for different types of contents is in everywhere.

Thanks again for the amazing Polymer project, I'm coding with it everyday on my engine and can't wait to show you guys the first demo. :)

**Comment by TimvdLippe** [comment]
The original issue seems to be fixed in your JSBin, therefore I am closing this issue.
